src/antidote_crdt_set_rw.erl

```diff
@@ -73,82 +73,123 @@
+  if
+    Elem1 == Elem2 ->
+      DownstreamOp = CreateDownstream(Elem1, AddTokens ++ RemoveTokens),
+      create_downstreams(CreateDownstream, ElemsRest, RWSetRest, [DownstreamOp|DownstreamOps]);
+    Elem1 > Elem2 ->
+      create_downstreams(CreateDownstream, Elems, RWSetRest, DownstreamOps);
+    true ->
+      DownstreamOp = CreateDownstream(Elem1, []),
+      create_downstreams(CreateDownstream, ElemsRest, RWSet, [DownstreamOp|DownstreamOps])
+  end.
 
 %% @doc generate a unique identifier (best-effort).
 unique() ->
     crypto:strong_rand_bytes(20).
 
-elems_with_removed_tombstones([], _Dict, _ResetTokens) -> [];
-elems_with_removed_tombstones(Elems, [], ResetTokens) -> [{E, ResetTokens} || E <- Elems];
-elems_with_removed_tombstones([E|ElemsRest]=Elems, [{K, Ts}|DictRest]=Dict, ResetTokens) ->
-    if
-        E == K ->
-            [{E, Ts ++ ResetTokens}|elems_with_removed_tombstones(ElemsRest, DictRest, ResetTokens)];
-        E > K ->
-            elems_with_removed_tombstones(Elems, DictRest, ResetTokens);
-        true ->
-            [{E, ResetTokens}|elems_with_removed_tombstones(ElemsRest, Dict, ResetTokens)]
-    end.
-
--spec update(effect(), set()) -> {ok, set()}.
-update({add, Elems, ObservedResetTokens}, {ResetTokens, Dict}) ->
-  NewResetTokens = ResetTokens -- ObservedResetTokens,
-  {ok, {ResetTokens, add_elems(Elems, Dict, [], NewResetTokens)}};
-update({remove, Elems, Token, ObservedResetTokens}, {ResetTokens, Dict}) ->
-  NewResetTokens = ResetTokens -- ObservedResetTokens,
-  {ok, {ResetTokens, add_elems(Elems, Dict, [Token], NewResetTokens)}};
-update({reset, Elems, OldResetTokens, NewResetToken}, {ResetTokens, Dict}) ->
-  Dict2 = add_elems(Elems, Dict, [], []),
-  Dict3 = [{X, [NewResetToken|Tokens -- OldResetTokens]} || {X, Tokens} <- Dict2],
-  {ok, {[NewResetToken] ++ (ResetTokens -- OldResetTokens), Dict3}}.
-
-
-add_elems([], Dict, _, _) -> Dict;
-add_elems(Elems, [], NewTombs, NewEntryTombs) -> [{E, NewTombs ++ NewEntryTombs} || {E, _} <- Elems];
-add_elems([{E, RemovedTombstones}|ElemsRest]=Elems, [{K, Ts}|DictRest]=Dict, NewTombs, NewEntryTombs) ->
-    if
-        E == K ->
-            [{E, NewTombs ++ (Ts -- RemovedTombstones)}|add_elems(ElemsRest, DictRest, NewTombs, NewEntryTombs)];
-        E > K ->
-            [{K, Ts}|add_elems(Elems, DictRest, NewTombs, NewEntryTombs)];
-        true ->
-            [{E, NewTombs ++ NewEntryTombs}|add_elems(ElemsRest, Dict, NewTombs, NewEntryTombs)]
-    end.
+-spec update(downstream_op(), set()) -> {ok, set()}.
+  update(DownstreamOp, RWSet) ->
+    {ok, apply_downstreams(DownstreamOp, RWSet)}.
+
+%% @private apply a list of downstream ops to a given orset
+apply_downstreams([], RWSet) ->
+  RWSet;
+apply_downstreams(Ops, []) ->
+  lists:foldl(
+    fun({Elem, SeenTokens, ToAdd, ToRemove}, RWSet) ->
+      RWSet ++ apply_downstream(Elem, SeenTokens, ToAdd, ToRemove, [], [])
+    end,
+    [],
+    Ops
+  );
```

Comment:
I think the foldl could be replaced by a map or a list comprehension: `[apply_downstream(Elem, SeenTokens, ToAdd, ToRemove, [], [])  || {Elem, SeenTokens, ToAdd, ToRemove} <- Ops]`